tests/miral/runner.cpp

```diff
@@ -35,6 +38,36 @@ struct Runner : miral::TestServer
     }
 
     MOCK_METHOD0(callback, void());
+
+    void wait_for_locking_callback()
+    {
+        std::lock_guard<std::mutex> lock(mutex);
+        ready_to_tear_down = false;
+    }
+
+    void locking_callback()
+    {
+        {
+            std::lock_guard<std::mutex> lock(mutex);
+            ready_to_tear_down = true;
+        }
+
+        cv.notify_one();
+        callback();
+    }
+
+    void wait_to_tear_down()
+    {
+        std::unique_lock<std::mutex> lock(mutex);
+        cv.wait(lock, [this] { return ready_to_tear_down; });
+        lock.unlock();
+    }
+
+    std::mutex mutex;
+    std::condition_variable cv;
+    bool ready_to_tear_down = true;
+    mt::Pipe pipe;
+    char const data_to_write = 'a';
```

Comment:
Given the semantics here, this looks like a helper class:

```suggestion
class ScopeExitSemaphore
{
public:
   ScopeExitSemaphore() = default;

    void ~ScopeExitSemaphore()
    {
        semaphore.acquire();
    }
   
    void operator()()
    {
        semaphore.release();
    }
private:
    std::binary_semaphore semaphore{0};
};
```

Then the tests can be written:

```
TEST_F(Runner, register_signal_handler_before_setup_invokes_callback_after_setup)
{
    int const signum{SIGUSR1};

    ScopeExitSemaphore scope_exit_signal;
    register_signal_handler({signum}, [this, &](int){ callback(); scope_exit_signal(); });

    miral::TestServer::SetUp();

    EXPECT_CALL(*this, callback())
        .Times(AtLeast(1));

    kill(getpid(), signum);
}

Reply:
It is moot (because I'm happy with either approach) but neither approach allows "FdHandle to go out of scope and unregister the fd even if the semaphore was held".